### PR TITLE
CIT Varianter with constraints

### PR DESCRIPTION
**Pull request by richtja**
The new version of CIT varianter which support constraints for combinations.

Signed-off-by: Jan Richter <jarichte@redhat.com>

### CODE REVIEW

**Comment by clebergnu** [comment]
This pull request **fixes 1 alert** when merging 9f34d28732924bdd84fb41a7ba951a01c40d3f3b into 729bb0493086812c843b72057967c0e911753f9b - [view on LGTM.com](https://lgtm.com/projects/g/avocado-framework/avocado/rev/pr-660ae498563dfe78407258355839eeef4733e531)

**fixed alerts:**

* 1 for Redundant comparison

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by clebergnu** [comment]
This pull request **fixes 1 alert** when merging 3d58988b1bbfa6eb95387f5ee82f63eacf7748ce into 2ab14d7519ce7fa4bd9deb921e7048bb733be780 - [view on LGTM.com](https://lgtm.com/projects/g/avocado-framework/avocado/rev/pr-acd45b21d29bd487de00b7df99175c9f1f74745e)

**fixed alerts:**

* 1 for Redundant comparison

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            self.solver.clean_data_matrix(data_matrix, {"name": parameter, "value": value_choice})
+            row.append(value_choice)
+        return row
+
+
+def print_progress(interaction):
+    maximum = round(ITERATIONS_SIZE / 100)
+    perc = round(interaction / 100)
+    message = "Computing[" + "*" * (perc) + " " * (maximum - perc) + "]" + "."
+    sys.stdout.write('\r' + message)
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py]: This is a very rough version of the LOG suggestions I made before:

```diff
diff --git a/optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py b/optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py
index 8b5c40d8..9425c2f5 100644
--- a/optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py
+++ b/optional_plugins/varianter_cit/avocado_varianter_cit/Cit.py
@@ -1,23 +1,28 @@
-from __future__ import print_function
-
 import random
 import sys
+import logging
 
 from .Solver import Solver
 from .CombinationMatrix import CombinationMatrix
 
+from avocado.core.output import LOG_UI
+
+LOG = LOG_UI.getChild("Cit")
+LOG.setLevel(logging.INFO)
+
+
 ITERATIONS_SIZE = 600
 
 
 class Cit:
 
-    def __init__(self, input_data, t_value, constraints, debug=False):
+    def __init__(self, input_data, t_value, constraints):
         """
         Creation of CombinationMatrix from user input
+
         :param input_data: parameters from user
         :param t_value: size of one combination
         :param constraints: constraints of combinations
-        :param debug: if debug statements should be produced
         """
         self.data = input_data
         self.t_value = t_value
@@ -28,11 +33,12 @@ class Cit:
         # Combinations which do not match to the constraints are disabled
         self.solver.clean_hash_table(self.combination_matrix, t_value)
         self.final_matrix = []
-        self.debug = debug
 
     def final_matrix_init(self):
         """
-        Creation of the first solution. This solution is the start of searching for the best solution
+        Creation of the first solution. This solution is the start of
+        searching for the best solution
+
         :return: solution matrix (list(list))
         """
         self.final_matrix = [self.create_random_row_with_constraints()]
@@ -48,9 +54,11 @@ class Cit:
 
     def compute(self):
         """
-        Searching for the best solution. It creates one solution and from that, it tries to create smaller solution
-        This searching process is limited by ITERATIONS_SIZE. When ITERATIONS_SIZE is 0 the last found solution is
-        the best solution
+        Searching for the best solution. It creates one solution and
+        from that, it tries to create smaller solution This searching
+        process is limited by ITERATIONS_SIZE. When ITERATIONS_SIZE is
+        0 the last found solution is the best solution
+
         :return: The best solution
         """
         self.final_matrix = self.final_matrix_init()
@@ -63,21 +71,17 @@ class Cit:
                 delete_row = matrix.pop(random.randint(0, len(matrix) - 1))
                 self.combination_matrix.uncover_solution_row(delete_row)
                 deleted_rows.append(delete_row)
-            if self.debug:
-                print("I'm trying soulution with size " + str(len(matrix)) + " and " + str(iterations) + " iterations")
+            LOG.debug("I'm trying solution with size %d and %d iterations",
+                      len(matrix), iterations)
             matrix, is_better_solution = self.find_better_solution(iterations, matrix)
             if is_better_solution:
                 self.final_matrix = matrix[:]
                 deleted_rows = []
                 step_size *= 2
-                if self.debug:
-                    print("-----solution with size " + str(len(matrix)) + " was found-----")
-                    print()
+                LOG.debug("-----solution with size %d was found-----", len(matrix))
                 iterations = ITERATIONS_SIZE
             else:
-                if self.debug:
-                    print("-----solution with size " + str(len(matrix)) + " was not found-----")
-                    print()
+                LOG.debug("-----solution with size %d was found-----", len(matrix))
                 for i in range(step_size):
                     self.combination_matrix.cover_solution_row(deleted_rows[i])
                     matrix.append(deleted_rows[i])
@@ -86,20 +90,18 @@ class Cit:
                 else:
                     step_size = 0
 
-        if self.debug:
-            print("-----The best solution is-----")
         return self.final_matrix
 
     def find_better_solution(self, counter, matrix):
         """
         Changing the matrix to cover all combinations
+
         :param counter: maximum number of changes in the matrix
         :param matrix: matrix to be changed
         :return: new matrix and is changes have been successful?
         """
         while self.combination_matrix.total_uncovered != 0:
-            if self.debug:
-                print_progress(counter)
+            print_progress(counter)
             solution, row_index, _ = self.use_random_algorithm(matrix)
             self.combination_matrix.uncover_solution_row(matrix[row_index])
             self.combination_matrix.cover_solution_row(solution)
@@ -305,4 +307,4 @@ def print_progress(interaction):
     maximum = round(ITERATIONS_SIZE / 100)
     perc = round(interaction / 100)
     message = "Computing[" + "*" * (perc) + " " * (maximum - perc) + "]" + "."
-    sys.stdout.write('\r' + message)
+    LOG.debug(message)
diff --git a/optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py b/optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py
index 1917db17..3073c0bb 100644
--- a/optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py
+++ b/optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py
@@ -14,6 +14,7 @@
 
 import os
 import sys
+import logging
 
 from avocado.core import exit_codes
 from avocado.core import varianter
@@ -21,7 +22,7 @@ from avocado.core.output import LOG_UI
 from avocado.core.plugin_interfaces import CLI
 from avocado.core.plugin_interfaces import Varianter
 from avocado.core.tree import TreeNode
-from avocado_varianter_cit.Cit import Cit
+from avocado_varianter_cit.Cit import Cit, LOG
 from avocado_varianter_cit.Parser import Parser
 
 
@@ -49,8 +50,10 @@ class VarianterCitCLI(CLI):
                                    "%(default)s, maximum number is specific "
                                    "to parameter file content"))
 
+
     def run(self, args):
-        pass
+        if getattr(args, "varianter_debug", False):
+            LOG.setLevel(logging.DEBUG)
 
 
 class VarianterCit(Varianter):
```

The interesting part is that by using the `--debug` command line option that already exists, the debug messages are printed or not.

**Comment by richtja** [comment]
Thanks for the initial review, I created new version #3114  with requested changes. So I am closing this.

**Pull request closed without merging**